pkg/issuer/acme/dns/route53/route53_test.go

```diff
@@ -77,24 +78,170 @@ func TestNoCredentialsFromEnv(t *testing.T) {
 	assert.Error(t, err, "Expected error constructing DNSProvider with no credentials and not ambient")
 }
 
-func TestAmbientRegionFromEnv(t *testing.T) {
-	t.Setenv("AWS_REGION", "us-east-1")
-
-	_, ctx := ktesting.NewTestContext(t)
-	provider, err := NewDNSProvider(ctx, "", "", "", "", "", "", true, util.RecursiveNameservers, "cert-manager-test")
-	assert.NoError(t, err, "Expected no error constructing DNSProvider")
+type bitmask byte
 
-	assert.Equal(t, "us-east-1", provider.client.Options().Region, "Expected Region to be set from environment")
+func (haystack bitmask) Has(needle bitmask) bool {
+	return haystack&needle != 0
 }
 
-func TestNoRegionFromEnv(t *testing.T) {
-	t.Setenv("AWS_REGION", "us-east-1")
+// TestSessionProviderGetSessionRegion calls sessionProvider.GetSession with all
+// permutations of those inputs that influence how it selects an AWS region.
+// The desired region selection properties are documented alongside each
+// assertion.
+func TestSessionProviderGetSessionRegion(t *testing.T) {
+	const (
+		fakeAmbientRegion = "ambient-region-1"
+		fakeIssuerRegion  = "issuer-region-1"
+	)
 
-	_, ctx := ktesting.NewTestContext(t)
-	provider, err := NewDNSProvider(ctx, "marx", "swordfish", "", "", "", "", false, util.RecursiveNameservers, "cert-manager-test")
-	assert.NoError(t, err, "Expected no error constructing DNSProvider")
+	const (
+		allowAmbientCredentials bitmask = 1 << iota
+		setAmbientRegion
+		supplyAccessKey
+		supplyWebIdentity
+		supplyIssuerRegion
+	)
+	allFalse := bitmask(0)
+	allTrue := allowAmbientCredentials | setAmbientRegion | supplyAccessKey | supplyWebIdentity | supplyIssuerRegion
```

Comment:
Can we move the bitmap stuff after the testFunc function, so it is clear that it is only used in the loop?

Reply:
Sure. Done.